File: _stats_build/statistics/wr_median.rb
require_relative "abstract/round_metric"

class WrMedian < RoundMetric
  def initialize
    @title = "Median"
    @note = "Median: the middle value of all solves in a round. With DNFs, the median shifts to a higher-ranked valid solve."
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right, "Person" => :left, "Competition" => :left, "Date" => :left, "Details" => :left }
  end

  # NOTE: Median = 5次成绩排序后的中位数（第3位）
  # 有 DNF/DNS 时中位数往后移：1个无效→第4位，2个无效→第5位，3个及以上→无效
  def compute_metric(values, r)
    valid = values.select { |v| v > 0 }.sort
    invalid_count = values.count { |v| v <= 0 }
    case invalid_count
    when 0
      valid[2]  # 5个有效，取第3个
    when 1
      valid[2]  # 4个有效，中位数位置从第3升到第3（排序后第3个）
    when 2
      valid[2]  # 3个有效，取最大的
    else
      nil       # 3个及以上无效，无法计算
    end
  end
end

File: core/jobs/stats-build/src/statistics/wr_median.ts
// NOTE: Median——一轮中的中位数
import { RoundMetric } from '../core/round_metric.js';
import { EVENTS_WITH_AVERAGE } from '../core/events.js';

export class WrMedian extends RoundMetric {
  constructor() {
    super();
    this.title = 'Median';
    this.titleZh = '中位数';
    this.note = 'Median: the middle value of all solves in a round. With DNFs, the median shifts to a higher-ranked valid solve.';
    this.noteZh = '中位数：一轮中所有成绩的中间值。有 DNF 时中位数向有效成绩偏移。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  targetEvents() { return EVENTS_WITH_AVERAGE; }

  // NOTE: Mo3/Ao5 都是奇数次。DNF 视为比任意有效成绩更差，再取正中间一项。
  computeMetric(values: number[]): number | null {
    if (values.length < 3 || values.length % 2 === 0) return null;
    const sorted = values
      .map(v => v > 0 ? v : Number.POSITIVE_INFINITY)
      .sort((a, b) => a - b);
    const median = sorted[Math.floor(sorted.length / 2)];
    return Number.isFinite(median) ? median : null;
  }
}
